spring-cloud-common-security-config-web/src/main/java/org/springframework/cloud/common/security/OAuthSecurityConfiguration.java

```diff
@@ -106,272 +108,440 @@
+                .oidcUserService(this.oidcUserService);
+
+        if (opaqueTokenIntrospector != null) {
+            http.oauth2ResourceServer()
+                    .opaqueToken()
+                    .introspector(opaqueTokenIntrospector);
+        } else if (oAuth2ResourceServerProperties.getJwt().getJwkSetUri() != null) {
+            http.oauth2ResourceServer()
+                    .jwt()
+                    .jwtAuthenticationConverter(grantedAuthoritiesExtractor());
+        }
+
+        this.securityStateBean.setAuthenticationEnabled(true);
+    }
+
+    protected static String dashboard(AuthorizationProperties authorizationProperties, String path) {
+        return authorizationProperties.getDashboardUrl() + path;
+    }
+
+    private static String calculateDefaultProviderId(AuthorizationProperties authorizationProperties, OAuth2ClientProperties oauth2ClientProperties) {
+        if (authorizationProperties.getDefaultProviderId() != null) {
+            return authorizationProperties.getDefaultProviderId();
+        } else if (oauth2ClientProperties.getRegistration().size() == 1) {
+            return oauth2ClientProperties.getRegistration().entrySet().iterator().next().getKey();
+        } else if (oauth2ClientProperties.getRegistration().size() > 1
+                && !StringUtils.hasText(authorizationProperties.getDefaultProviderId())) {
+            throw new IllegalStateException("defaultProviderId must be set if more than 1 Registration is provided.");
+        } else {
+            throw new IllegalStateException("Unable to retrieve default provider id.");
+        }
+    }
+
+
+    protected Converter<Jwt, AbstractAuthenticationToken> grantedAuthoritiesExtractor() {
+        String providerId = calculateDefaultProviderId(authorizationProperties, oauth2ClientProperties);
+        ProviderRoleMapping providerRoleMapping = authorizationProperties.getProviderRoleMappings().get(providerId);
+
+        JwtAuthenticationConverter jwtAuthenticationConverter =
+                new JwtAuthenticationConverter();
+
+        MappingJwtGrantedAuthoritiesConverter converter = new MappingJwtGrantedAuthoritiesConverter();
+        converter.setAuthorityPrefix("");
+        jwtAuthenticationConverter.setJwtGrantedAuthoritiesConverter(converter);
+        if (providerRoleMapping != null) {
+            converter.setAuthoritiesMapping(providerRoleMapping.getRoleMappings());
+            converter.setGroupAuthoritiesMapping(providerRoleMapping.getGroupMappings());
+            if (StringUtils.hasText(providerRoleMapping.getPrincipalClaimName())) {
+                jwtAuthenticationConverter.setPrincipalClaimName(providerRoleMapping.getPrincipalClaimName());
+            }
+        }
+        return jwtAuthenticationConverter;
+    }
+
+    @Configuration
+    @ConditionalOnProperty(prefix = "spring.security.oauth2.resourceserver.opaquetoken", value = "introspection-uri")
+    public static class OpaqueTokenIntrospectorConfig {
```

Comment:
I love the technique of wrapping the beans in configs as they behave much better during auto-config wrt to conditionals etc..

1. Lets set the configs to `proxyBeanMethods = true` 
2. Lets move the access to package-protected as they were not visible before. Otherwise we have to consider it public API etc..
3. Is there any existing tests that exercise the enabled/disabled flags that are guarding these configs? Hopefully there were already tests in place for that.  

(comment applies to all newly added configs in here).

Reply:
Doesn't `proxyBeanMethods = true`  will need to turn to false to be ready for Spring Native.

Reply:
I totally meant `proxyBeanMethods = "false"` as `true` is already the default.